### ISSUE TITLE
[WIP] [Benchmark] HF Trainer on A100

### ISSUE THREAD

**Issue by stas00** [huggingface/transformers#15026]
# 🖥 Benchmarking RTX-3090 `transformers` w/ HF Trainer:

We are going to use a special benchmarking tool that will do all the work for us. https://github.com/huggingface/transformers/pull/14934

This is the index post and specific benchmarks are in their own posts below:

Please ignore for now as I'm gathering results.

**Comment by stas00** [comment]
# precision: fp16 vs bf16 vs tf32 vs fp32

Main interest: benchmarking the new --bf16 and --tf32 on Ampere/RTX-3090, comparatively to fp16 and fp32 modes.

- bf16 is `autocast(dtype=torch.bfloat16)`
- tf32 is `torch.backends.cuda.matmul.allow_tf32 = True`



## Benchmark

The benchmark uses 3 different t5 models, and at the end of the section also gpt2. For t5 the main script is:

```

```

and now adding one of:

```
--tf32 0 # fp32
--tf32 0 --fp16
--tf32 0 --bf16
--tf32 1
--tf32 1 --fp16
--tf32 1 --bf16
```

But we are going to use a special benchmarking tool that will do all the work for us. https://github.com/huggingface/transformers/pull/14934

Important notes:
1. `--tf32 0 --fp16 0` combo is just fp32 (which is the default mode - we don't have this option per se)
2. I changed `--per_device_train_batch_size` in the base command from 32 (`t5-small`) to 16  (`t5-base`) to 8 (`t5-large`) to be able to fit into the GPU memory while keeping it as occupied as possible.
3. I changed `--max_train_samples` in the base command from 20k (`t5-small`) to 10k  (`t5-base`) to 5k (`t5-large`) to give each run about 1-3min of run time so that the benchmark doesn't take too too long, but is long enough to put strain on the card.


*** Setup:

```
Datetime    : 2022-01-03 22:43:38

Software:
transformers: 4.16.0.dev0
torch       : 1.10.1
cuda        : 11.3
python      : 3.8.12

Hardware:
1 GPUs      : NVIDIA A100-SXM4-40GB, 39.59GB
```

## Benchmark 4: gpt2

*** Results:

| Variation       |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------|------------------------------------:|------------:|----------------:|
| --tf32 0        |                               28.77 |         100 |            3.34 |
| --tf32 1        |                               63.51 |         221 |            3.34 |
| --fp16 --tf32 0 |                               69.60 |         242 |            3.34 |
| --fp16 --tf32 1 |                               69.98 |         243 |            3.34 |
| --bf16 --tf32 0 |                               70.37 |         245 |            3.34 |
| --bf16 --tf32 1 |                               69.88 |         243 |            3.34 |

```
*** The benchmark command line was:

CUDA_VISIBLE_DEVICES=0 python ./scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/language-modeling/run_clm.py --model_name_or_path gpt2 \
--dataset_name wikitext --dataset_config_name wikitext-2-raw-v1 \
--logging_strategy no --save_strategy no --do_train --max_train_samples 5000 \
--per_device_train_batch_size 16 --num_train_epochs 1 --warmup_steps 8 \
--block_size 512 --report_to none ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'|--fp16|--bf16' '--tf32 0|--tf32 1' --report-metric-keys train_loss \
--repeat-times 1 --base-variation '--tf32 0'

```

**Comment by stas00** [comment]
# gradient accumulation steps

Let's choose `t5-base` model to test with as it's pretty large yet doesn't overflow like t5-large.

Let's measure `--gradient_accumulation_steps` 1,2,4,8,16,32 with different precision configurations.

| Variation                                        |     Train |   Diff |   Train |
|                                                  |   samples |      % |    loss |
|                                                  |       per |        |         |
|                                                  |    second |        |         |
|:-------------------------------------------------|----------:|-------:|--------:|
| --gradient_accumulation_steps 1 --tf32 0         |     93.68 |    100 |    2.21 |
| --gradient_accumulation_steps 1 --tf32 1         |    210.53 |    225 |    2.21 |
| --gradient_accumulation_steps 1 --tf32 0 --fp16  |    217.75 |    232 |    2.21 |
| --gradient_accumulation_steps 1 --tf32 0 --bf16  |    224.09 |    239 |    2.22 |
| --gradient_accumulation_steps 2 --tf32 0         |     97.48 |    104 |    2.28 |
| --gradient_accumulation_steps 2 --tf32 1         |    236.39 |    252 |    2.28 |
| --gradient_accumulation_steps 2 --tf32 0 --fp16  |    244.81 |    261 |    2.28 |
| --gradient_accumulation_steps 2 --tf32 0 --bf16  |    246.08 |    263 |    2.29 |
| --gradient_accumulation_steps 4 --tf32 0         |     99.68 |    106 |    2.39 |
| --gradient_accumulation_steps 4 --tf32 1         |    248.24 |    265 |    2.40 |
| --gradient_accumulation_steps 4 --tf32 0 --fp16  |    259.20 |    277 |    2.41 |
| --gradient_accumulation_steps 4 --tf32 0 --bf16  |    263.39 |    281 |    2.42 |
| --gradient_accumulation_steps 8 --tf32 0         |    100.67 |    107 |    2.58 |
| --gradient_accumulation_steps 8 --tf32 1         |    252.45 |    269 |    2.58 |
| --gradient_accumulation_steps 8 --tf32 0 --fp16  |    261.59 |    279 |    2.58 |
| --gradient_accumulation_steps 8 --tf32 0 --bf16  |    267.37 |    285 |    2.62 |
| --gradient_accumulation_steps 16 --tf32 0        |    100.97 |    108 |    2.83 |
| --gradient_accumulation_steps 16 --tf32 1        |    253.68 |    271 |    2.84 |
| --gradient_accumulation_steps 16 --tf32 0 --fp16 |    256.13 |    273 |    2.84 |
| --gradient_accumulation_steps 16 --tf32 0 --bf16 |    274.14 |    293 |    2.89 |



| Variation                        |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:---------------------------------|------------------------------------:|------------:|----------------:|
| --gradient_accumulation_steps 1  |                              221.72 |         100 |            2.22 |
| --gradient_accumulation_steps 2  |                              241.67 |         109 |            2.29 |
| --gradient_accumulation_steps 4  |                              252.79 |         114 |            2.42 |
| --gradient_accumulation_steps 8  |                              265.06 |         120 |            2.62 |
| --gradient_accumulation_steps 16 |                              265.36 |         120 |            2.89 |


```

1.

CUDA_VISIBLE_DEVICES=3 python ./scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/translation/run_translation.py --model_name_or_path t5-base \
--output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no \
--save_strategy no --per_device_train_batch_size 32 --max_source_length 512 \
--max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: " --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--gradient_accumulation_steps 1|--gradient_accumulation_steps 2|--gradient_accumulation_steps 4|--gradient_accumulation_steps 8|--gradient_accumulation_steps 16' \
'--tf32 0|--tf32 1|--tf32 0 --fp16|--tf32 0 --bf16' --report-metric-keys \
train_loss --repeat-times 1 

2.

CUDA_VISIBLE_DEVICES=3 python ./scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/translation/run_translation.py --model_name_or_path t5-base \
--output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no \
--save_strategy no --per_device_train_batch_size 32 --max_source_length 512 \
--max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: " --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 --tf32 0 --bf16 ' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--gradient_accumulation_steps 1|--gradient_accumulation_steps 2|--gradient_accumulation_steps 4|--gradient_accumulation_steps 8|--gradient_accumulation_steps 16' \
--report-metric-keys train_loss --repeat-times 1
```

**Comment by stas00** [comment]
# batch size

| Variation                        |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:---------------------------------|------------------------------------:|------------:|----------------:|
| --per_device_train_batch_size 1  |                                7.77 |         100 |            1.90 |
| --per_device_train_batch_size 2  |                               15.51 |         200 |            2.01 |
| --per_device_train_batch_size 4  |                               29.66 |         382 |            2.09 |
| --per_device_train_batch_size 8  |                               61.16 |         787 |            2.16 |
| --per_device_train_batch_size 16 |                              115.84 |        1492 |            2.25 |
| --per_device_train_batch_size 32 |                              224.96 |        2897 |            2.38 |

```

CUDA_VISIBLE_DEVICES=3 python ./scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/translation/run_translation.py --model_name_or_path t5-base \
--output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no \
--save_strategy no --max_source_length 512 \
--max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: " --warmup_steps 50 \
--max_train_samples 5000 --dataloader_num_workers 2 --bf16' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--per_device_train_batch_size 1|--per_device_train_batch_size 2|--per_device_train_batch_size 4|--per_device_train_batch_size 8|--per_device_train_batch_size 16|--per_device_train_batch_size 32' \
--report-metric-keys train_loss --repeat-times 1 

```

**Comment by stas00** [comment]
# gradient checkpointing

| Variation                  |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:---------------------------|------------------------------------:|------------:|----------------:|
| --gradient_checkpointing 0 |                              225.67 |         124 |            2.30 |
| --gradient_checkpointing 1 |                              182.68 |         100 |            2.30 |

```
CUDA_VISIBLE_DEVICES=3 python ./scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' examples/pytorch/translation/run_translation.py --model_name_or_path t5-base \
--output_dir output_dir --do_train --label_smoothing 0.1 --logging_strategy no \
--save_strategy no --per_device_train_batch_size 32 --max_source_length 512 \
--max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: " --warmup_steps 50 \
--max_train_samples 10000 --dataloader_num_workers 2 --bf16' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--gradient_checkpointing 0|--gradient_checkpointing 1' --report-metric-keys \
train_loss --repeat-times 1 

```

**Comment by stas00** [comment]
# optimizers

(Note: requires this PR to be merged first: https://github.com/huggingface/transformers/pull/14744)

| Variation               |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:------------------------|------------------------------------:|------------:|----------------:|
| --optim adam_hf         |                              214.55 |           2 |            2.21 |
| --optim adam_torch      |                              209.72 |           0 |            2.21 |
| --optim adafactor       |                              158.56 |         -24 |            2.21 |
| --optim adam_apex_fused |                              227.96 |           9 |            2.21 |

Observations:
- apex's FusedAdam is the fastest.

```
CUDA_VISIBLE_DEVICES=3 python scripts/benchmark/trainer-benchmark.py --base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adam_hf|--optim adam_torch|--optim adafactor|--optim adam_apex_fused' \
--report-metric-keys train_loss --base-variation '--optim adam_torch'
```

**Comment by stas00** [comment]
# speed up combinations

Now let's combine the winning strategies from each individual benchmark above and compare with the baseline:

| Variation                                                             |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------------------------------------------------------------|------------------------------------:|------------:|----------------:|
| --optim adam_torch --gradient_accumulation_steps 1 --tf32 0           |                               92.15 |           0 |            2.21 |
| --optim adam_apex_fused --gradient_accumulation_steps 8 --tf32 --bf16 |                              267.21 |         190 |            2.62 |

Getting an almost 3x improvement in speed!

```
CUDA_VISIBLE_DEVICES=3 python \
../transformers-stas/scripts/benchmark/trainer-benchmark.py --base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 32 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adam_torch --gradient_accumulation_steps 1 --tf32 0|--optim adam_apex_fused --gradient_accumulation_steps 8 --tf32 --bf16' \
--report-metric-keys train_loss --base-variation '--optim adam_torch'
```

**Comment by stas00** [comment]
# RTX-3090 vs A100

In all the benchmarks above I was making the batch size bigger and run more samples comparative to the same [RTX-3090 benchmarks](https://github.com/huggingface/transformers/issues/14608) as A100 40GB card can handle more than RTX-3090 24GB, but let's compare now the 2 using the same config. So we will have RTX-3090 fully loaded, but A100 will be only partially loaded.

Also each card is running on a different machines so there is a bit of hardware difference as well.



**A100**

| Variation                                                             |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------------------------------------------------------------|------------------------------------:|------------:|----------------:|
| --optim adam_torch --gradient_accumulation_steps 1 --tf32 0           |                               85.99 |           0 |            2.16 |
| --optim adam_apex_fused --gradient_accumulation_steps 8 --tf32 --bf16 |                              153.72 |          79 |            2.42 |


**RTX-3090**

| Variation                                                             |   Train<br>samples<br>per<br>second |   Diff<br>% |   Train<br>loss |
|:----------------------------------------------------------------------|------------------------------------:|------------:|----------------:|
| --optim adam_torch --gradient_accumulation_steps 1 --tf32 0           |                               88.94 |           0 |            2.16 |
| --optim adam_apex_fused --gradient_accumulation_steps 8 --tf32 --bf16 |                              173.15 |          95 |            2.42 |


Observations:

- This is very unexpected, RTX-3090 appears to be faster when we use only 1/2 of A100's capacity to match the 2.
- as the machines aren't the same it'd be good to find a machine that has both cards and test them on equal hardware.

```
CUDA_VISIBLE_DEVICES=0 python \
/hf/transformers-trainer-benchmark/scripts/benchmark/trainer-benchmark.py \
--base-cmd \
' \
examples/pytorch/translation/run_translation.py --model_name_or_path t5-base --output_dir output_dir \
--do_train --label_smoothing 0.1 --logging_strategy no --save_strategy no --per_device_train_batch_size 16 \
--max_source_length 512 --max_target_length 512 --num_train_epochs 1 --overwrite_output_dir \
--source_lang en --target_lang ro --dataset_name wmt16 --dataset_config "ro-en" \
--source_prefix "translate English to Romanian: "  --warmup_steps 50 \
--max_train_samples 20000 --dataloader_num_workers 2 \
' \
--target-metric-key train_samples_per_second --repeat-times 1 --variations \
'--optim adam_torch --gradient_accumulation_steps 1 --tf32 0|--optim adam_apex_fused --gradient_accumulation_steps 8 --tf32 --bf16' \
--report-metric-keys train_loss

```

**Comment by eqy** [comment]
@stas00 Another interesting issue that I found regarding batch size is that it is an important parameter when the model is mostly in fp32 and relies on autocast to dispatch to fp16 or bf16. I believe this is because of the overhead of casting back-and-forth can dominate the total runtime compared to the actual kernel/operator.

Consider the following microbenchmark:

```
import time
import torch

from torch.cuda.amp import autocast

def bench(dtype, dim, shape, auto=True):
    linear = torch.nn.Linear(shape[-1], dim, device='cuda')
    inp = torch.randn(shape, device='cuda')
    ctx_manager = None
    if not auto:                                                                                                                                                                                                                                     inp = inp.to(dtype)                                                                                                                                                                                                                          linear = linear.to(dtype)
    else:                                                                                                                                                                                                                                            ctx_manager = autocast(dtype=dtype)
                                                                                                                                                                                                                                                 def run(inp, layer, ctx_manager):                                                                                                                                                                                                                if ctx_manager is not None:                                                                                                                                                                                                                      with ctx_manager:                                                                                                                                                                                                                                layer(inp)                                                                                                                                                                                                                           else:                                                                                                                                                                                                                                            layer(inp)
                                                                                                                                                                                                                                                 run(inp, linear, ctx_manager)
    torch.cuda.synchronize()
    t1 = time.time()
    for i in range(1000):                                                                                                                                                                                                                            run(inp, linear, ctx_manager)
    torch.cuda.synchronize()
    t2 = time.time()
    return t2 - t1

if __name__ == '__main__':
    hidden_dim = 1024
    for auto in (True, False):                                                                                                                                                                                                                       print(f"autocast: {auto}")                                                                                                                                                                                                                   for batch_size in (8, 16, 32, 64):                                                                                                                                                                                                               shape = [batch_size, 56, hidden_dim]                                                                                                                                                                                                         times = list()                                                                                                                                                                                                                               for dtype in (torch.float32, torch.float16, torch.bfloat16):                                                                                                                                                                                     times.append(bench(dtype, hidden_dim, shape, auto))                                                                                                                                                                                      print(f"batch_size: {batch_size} fp32 {times[0]:3f} fp16 {times[1]:3f} bf16{times[2]:3f}")                                                                                                                                                   print(f"speedup: {batch_size} fp32 {(times[0]/times[0]):3f} fp16 {(times[0]/times[1]):3f} bf16 {(times[0]/times[2]):3f}")
```
I get the following times on A6000 (similar architecture to 3090):
```
autocast: True
batch_size: 8 fp32 0.041687 fp16 0.062801 bf160.062099
speedup: 8 fp32 1.000000 fp16 0.663797 bf16 0.671297
batch_size: 16 fp32 0.051690 fp16 0.063733 bf160.063265
speedup: 16 fp32 1.000000 fp16 0.811033 bf16 0.817039
batch_size: 32 fp32 0.112580 fp16 0.090171 bf160.092151
speedup: 32 fp32 1.000000 fp16 1.248507 bf16 1.221684
batch_size: 64 fp32 0.190485 fp16 0.152158 bf160.151925
speedup: 64 fp32 1.000000 fp16 1.251896 bf16 1.253812
autocast: False
batch_size: 8 fp32 0.039313 fp16 0.032014 bf160.031317
speedup: 8 fp32 1.000000 fp16 1.227984 bf16 1.255320
batch_size: 16 fp32 0.050257 fp16 0.032606 bf160.032005
speedup: 16 fp32 1.000000 fp16 1.541342 bf16 1.570312
batch_size: 32 fp32 0.113147 fp16 0.057424 bf160.060994
speedup: 32 fp32 1.000000 fp16 1.970368 bf16 1.855040
batch_size: 64 fp32 0.192178 fp16 0.104193 bf160.105336
speedup: 64 fp32 1.000000 fp16 1.844435 bf16 1.824421
```

**Comment by stas00** [comment]
Thank you, @eqy. Before trying to understand the results, isn't your benchmark missing casting the results back for non-amp case? Added it.

Also added a nicely formatted output so it's much easier to make sense of.

Please have a look at the modified version:

[bench.txt](https://github.com/huggingface/transformers/files/7873005/bench.txt)

**Comment by eqy** [comment]
I believe there are "speed-of-light" cases where the cast-back wouldn't be necessary, though this may not be possible for the architectures we're interested in. Here, I think the big picture is that once the batch-size falls below a certain amount, the "building-block" operations like GEMMs will be slower in reduced precision vs. fp32 when casts are needed.

**Comment by stas00** [comment]
why do you think bs=32 is an oddball relative to other bs for speedup? in both cases w/ and w/o amp its relatively faster for bf16 and fp16 then bs=64, and much more significantly for fp16. One would expect 8 < 16 < 32 < 64, but here it is 8 < 16 < 64< 32,

**Comment by eqy** [comment]
That's interesting, I didn't see quite so dramatic results on an A100 (80GB), 2 runs:

```
## Autocast: True


Results:
|   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |
|    8 |           0.062 |           0.086 |            0.077 |
|   16 |           0.043 |           0.089 |            0.085 |
|   32 |           0.073 |           0.084 |            0.084 |
|   64 |           0.119 |           0.101 |            0.112 |

Speedup:
|   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |
|    8 |           1.000 |           0.714 |            0.805 |
|   16 |           1.000 |           0.486 |            0.508 |
|   32 |           1.000 |           0.865 |            0.871 |
|   64 |           1.000 |           1.181 |            1.058 |                                                                                                                                                                                                                                                                                                                                                                                                                           ## Autocast: False                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Results:                                                                                                                                                                                                                                     |   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |                                                                                                                                                                                                                                                                                                                                                          |    8 |           0.045 |           0.049 |            0.040 |                                                                                                                                                                              |   16 |           0.041 |           0.048 |            0.047 |                                                                                                                                                                              |   32 |           0.073 |           0.046 |            0.044 |                                                                                                                                                                              |   64 |           0.120 |           0.063 |            0.076 |                                                                                                                                                                                                                                                                                                                                                                                                                           Speedup:                                                                                                                                                                                                                                     |   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |                                                                                                                                                                              |-----:|----------------:|----------------:|-----------------:|                                                                                                                                                                              |    8 |           1.000 |           0.908 |            1.129 |                                                                                                                                                                              |   16 |           1.000 |           0.855 |            0.873 |                                                                                                                                                                              |   32 |           1.000 |           1.570 |            1.638 |                                                                                                                                                                              |   64 |           1.000 |           1.913 |            1.580 | 

## Autocast: False


Results:
|   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |
|    8 |           0.044 |           0.048 |            0.041 |
|   16 |           0.041 |           0.047 |            0.048 |
|   32 |           0.073 |           0.045 |            0.047 |
|   64 |           0.120 |           0.063 |            0.077 |

Speedup:
|   bs |   torch.float32 |   torch.float16 |   torch.bfloat16 |
|    8 |           1.000 |           0.929 |            1.094 |
|   16 |           1.000 |           0.883 |            0.861 |
|   32 |           1.000 |           1.615 |            1.541 |
|   64 |           1.000 |           1.907 |            1.562 |
```

**Comment by pratikchhapolika** [comment]
> # 🖥 Benchmarking `transformers` w/ HF Trainer on A100 40GB
> We are going to use a special benchmarking tool that will do all the work for us. #14934
> 
> This is the index post and specific benchmarks are in their own posts below:
> 
> This is the index post and specific benchmarks are in their own posts below:
> 
> 1. [fp16 vs bf16 vs tf32 vs fp32](https://github.com/huggingface/transformers/issues/15026#issuecomment-1004543189)
> 2. [gradient accumulation steps](https://github.com/huggingface/transformers/issues/15026#issuecomment-1004592231)
> 3. [batch size](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005033957)
> 4. [gradient checkpointing](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005034578)
> 5. [optimizers](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005220263)
> 6. [combining winning strategies](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005227577) **~3x speed improvement!**
> 7. [RTX-3090 vs A100](https://github.com/huggingface/transformers/issues/15026#issuecomment-1005235845)
> 
> Note that each benchmark was run only once, so multiple runs and averaging is probably going to give slightly different results. The purpose here though is to see relative differences roughly and not try to give an exact number.
> 
> See also the [same benchmarks for RTX-3090](https://github.com/huggingface/transformers/issues/14608)

Is all benchmarking done on  A100 ("NVIDIA_TESLA_A100") single GPU? Can you also include CUDA memory required Vs Data points for training Vs No. of GPU's.

**Comment by stas00** [comment]
> Is all benchmarking done on A100 ("NVIDIA_TESLA_A100") single GPU? 

Yes.

> Can you also include CUDA memory required Vs Data points for training Vs No. of GPU's.

I don't understand your question.

**Comment by pratikchhapolika** [comment]
> adamw_hf

On how many data points and epochs is it benchmarked on with Single GPU?

**Comment by pratikchhapolika** [comment]
> > Is all benchmarking done on A100 ("NVIDIA_TESLA_A100") single GPU?
> 
> Yes.
> 
> > Can you also include CUDA memory required Vs Data points for training Vs No. of GPU's.
> 
> I don't understand your question.

On how many data points and epochs is it benchmarked on with Single GPU?

**Comment by pratikchhapolika** [comment]
> > > Is all benchmarking done on A100 ("NVIDIA_TESLA_A100") single GPU?
> > 
> > 
> > Yes.
> > > Can you also include CUDA memory required Vs Data points for training Vs No. of GPU's.
> > 
> > 
> > I don't understand your question.
> 
> On how many data points and epochs is it benchmarked on with Single GPU?

I get error with 4 GPU's, 20 epochs on A100 with 700000 data points

`python -m torch.distributed.launch --nproc_per_node 4 train.py --gradient_accumulation_steps 8 --per_device_train_batch_size 8 --optim adamw_hf --tf32 --bf16"])`




`'Traceback (most recent call last):\n', ' File "train.py", line 160, in <module>\n trainer.train()\n', ' File "/opt/conda/lib/python3.7/site-packages/transformers/trainer.py", line 1398, in train\n tr_loss_step = self.training_step(model, inputs)\n', ' File "/opt/conda/lib/python3.7/site-packages/transformers/trainer.py", line 1994, in training_step\n self.scaler.scale(loss).backward()\n', ' File "/opt/conda/lib/python3.7/site-packages/torch/_tensor.py", line 363, in backward\n torch.autograd.backward(self, gradient, retain_graph, create_graph, inputs=inputs)\n', ' File "/opt/conda/lib/python3.7/site-packages/torch/autograd/__init__.py", line 175, in backward\n allow_unreachable=True, accumulate_grad=True) # Calls into the C++ engine to run the backward pass\n', 'RuntimeError: CUDA out of memory. Tried to allocate 3.82 GiB (GPU 0; 39.59 GiB total capacity; 17.07 GiB already allocated; 2.75 GiB free; 21.43 GiB reserved in total by PyTorch) If reserved memory is >> allocated memory try setting max_split_size_mb to avoid fragmentation. See documentation for Memory Management and PYTORCH_CUDA_ALLOC_CONF\n'`

**Comment by stas00** [comment]
> On how many data points and epochs is it benchmarked on with Single GPU?

It's defined by `--max_train_samples`

To your last question - please let's not derail this Benchmark Issue. If want to discuss an unrelated question please open a new issue. Best to delete it from here and post in another Issue.
